pydough/conversion/hybrid_tree.py

```diff
@@ -605,6 +623,97 @@ def add_child(
 
+    @staticmethod
+    def infer_metadata_reverse_cardinality(
+        metadata: SubcollectionRelationshipMetadata,
+    ) -> JoinCardinality:
+        """
+        Infers the cardinality of the reverse of a join from parent to child
+        based on the metadata of the reverse-relationship, if one exists.
+
+        Args:
+            `metadata`: the metadata for the sub-collection property mapping
+            the parent to the child.
+
+        Returns:
+            The join cardinality for the connection from the child back to the
+            parent, if it can be inferred. Uses `PLURAL_FILTER` as a fallback.
+        """
+        # If there is no reverse, fall back to plural filter (which is the
+        # safest default assumption).
+        if (
+            not isinstance(metadata, ReversiblePropertyMetadata)
+            or metadata.reverse is None
+        ):
+            return JoinCardinality.PLURAL_FILTER
+
+        # If the reverse property exists, use its properties to
+        # infer if the reverse cardinality is singular or plural
+        # and whether a match always exists or not.
+        cardinality: JoinCardinality
+        match (metadata.reverse.is_plural, metadata.reverse.always_matches):
+            case (False, True):
+                cardinality = JoinCardinality.SINGULAR_ACCESS
+            case (False, False):
+                cardinality = JoinCardinality.SINGULAR_FILTER
+            case (True, True):
+                cardinality = JoinCardinality.PLURAL_ACCESS
+            case (True, False):
+                cardinality = JoinCardinality.PLURAL_FILTER
+        return cardinality
+
+    def infer_root_reverse_cardinality(self, context: "HybridTree") -> JoinCardinality:
+        """
+        Infers the cardinality of the join connecting the root of the hybrid
+        tree to its parent context.
+
+        Args:
+            `context`: the parent context that the root of the hybrid tree is
+            being connected to.
+
+        Returns:
+            The inferred cardinality of the join connecting the root of the
+            hybrid tree to its parent context.
+        """
+        # Keep traversing upward until we find the root of the current tree.
+        if self.parent is not None:
+            return self.parent.infer_root_reverse_cardinality(context)
+
+        # Once we find the root, infer the cardinality of the join that would
+        # connect just this node to the parent context. The rest of the nodes in
+        # the tree don't matter since they will not affect how many matches
+        # there are back to the parent context, or whether there is always a
+        # match or not for each record in the current context.
+        match self.pipeline[0]:
+            case HybridRoot():
+                # If the parent of the child is a root, it means a cross join
+                # is occurring, so the cardinality depends on the size of the
+                # parent.
```

Comment:
not clear to me what `size` here means.
```suggestion
                # is occurring, so the cardinality depends on whether
                # the parent context is singular or plural.
```